3700bridge

```diff
@@ -84,140 +153,235 @@ class Bridge(object):
-        self.routing_table = dict()
 
-    def reset_routing_table(self):
-        self.routing_table = dict()
+        self.last_bpdu_send = 0
+        self.bpdus = {port.num: [] for port in self.ports}
 
-    def update_root(self, root_bridge: str, root_port: int, root_cost: int):
-        self.root_bridge = root_bridge
-        self.reset_routing_table()
-        print('New root: {}/{}'.format(serialize_address(self.bridge_id), root_bridge))
+        self.next_bridge = None
+        self.root_port = None
+        self.root_cost = 0
+        self.update_root(self.bridge_id, None, None, 0)
 
-        self.update_root_port(root_port, root_cost)
-
-    def update_root_port(self, root_port: int, root_cost: int):
-        self.root_port = root_port
-        self.root_cost = root_cost
-        print('Root port: {}/{}'.format(serialize_address(self.bridge_id), root_port))
+        self.forwarding_table = ForwardingTable()
 
     def run(self):
-        last_cycle = datetime.utcnow()
         while True:
-            since_last_run = (datetime.utcnow() - last_cycle).total_seconds()
-            if (since_last_run < MIN_THREAD_SLEEP):
-                sleep(MIN_THREAD_SLEEP - since_last_run)
-            last_cycle = datetime.utcnow()
-
             packets = self.read_packets()
-            bpdus = filter_bpdus(packets)
-            data = filter_data(packets)
+            self.process_packets(packets)
+
+            if VERBOSE and VERBOSE_MST:
+                print('Ports: {}'.format(
+                    [port.lan for port in self.ports if port.is_open]))
 
-            self.update_routing_table_ttls()
-            self.process_bpdus(bpdus)
-            for packet in data:
-                self.process_data_packet(packet)
             self.send_bpdus()
 
-    def process_data_packet(self, packet):
-        print('Received message {}, on port {} from {} to {}').format(
-            packet.id(), packet.port, packet.source, packet.dest)
-        self.update_routing_table(packet)
-        self.process_packet(packet)
-
-    def update_routing_table(self, packet):
-        self.routing_table[packet.source] = RoutingEntry(packet.port)
-
-    def update_routing_table_ttls(self):
-        for bridge, entry in self.routing_table:
-            if (datetime.utcnow() - entry.created_at).total_seconds() > 5:
-                self.routing_table.pop(bridge, False)
-
-    def process_packet(self, packet):
-        if self.routing_table.get(packet.dest):
-            port = self.ports.get(self.routing_table[packet.dest].port)
-            if port and port.is_open:
-                self.forward_packet(packet, port)
-            else:
-                self.drop_packet(packet)
-        else:
-            self.broadcast_packet(packet)
+    def read_packets(self):
+        ready_ports, _, _ = select(self.ports, [], [], SELECT_TIMEOUT)
+        packets = [port.recv_packet() for port in ready_ports]
+        return [packet for packet in packets if packet]
```

Comment:
I just realized how stupid this line sounds but it's filtering out any `None`s that `recv_packet` returned.